README.md
# northstar-cloud

A microservice for providing NorthStar cloud operations.

### Design
Requirements


### Developer Setup
Tools: Python, gRPC framework, REST framework, MySQL, Docker and Kubernetes.

First install some basic Python things.

if linux,
sudo apt install python python-dev python-pip
sudo pip install -U tox


Now, Go to the root of this repo (northstar-cloud)

To install northstar-cloud,
``` shell
cd northstar-cloud/
pip install -r requirements.txt
python setup.py install
```

To test everything OK, run tox (automation testing library).

``` shell
tox -epy36
```

add user into northstar-cloud from cli,
``` shell
python northstar_cloud/clients/northstart_service.py examples/sample-user.json

2019-06-20 15:39:31,373 - __main__ - INFO - northstar-service-client: Request :user {
  user_name: "help.me"
  first_name: "acb"
  last_name: "xyz"
  phone_number: "4089779890"
  home_address: "IBM Silicon valley"
  email_address: "abc@xyz.com"
  office_address: "IBM Silicon valley"
  app_id: "app_id_1"
  app_type: "iPhone"
  current_location {
    latitude: 37.121
    longitude: -121.5655
  }
  health_info {
    need_medical_support: true
  }
}

2019-06-20 15:39:31,697 - __main__ - INFO - northstar-service-client: Response :success: true
```

Run server in one terminal,
``` shell
python northstar_cloud/cli/northstar_cloud_start.py 

2019-06-20 15:39:20,074 - __main__ - INFO - northstar-cloud: service stating...
2019-06-20 15:39:20,076 - __main__ - INFO - northstar-cloud: is runnnig at ... localhost:50051
```

Run client in other terminal,
``` shell
python northstar_cloud/clients/northstart_service.py examples/sample-user.json

2019-06-20 15:39:31,373 - __main__ - INFO - northstar-service-client: Request :user {
  user_name: "help.me"
  first_name: "acb"
  last_name: "xyz"
  phone_number: "4089779890"
  home_address: "IBM Silicon valley"
  email_address: "abc@xyz.com"
  office_address: "IBM Silicon valley"
  app_id: "app_id_1"
  app_type: "iPhone"
  current_location {
    latitude: 37.121
    longitude: -121.5655
  }
  health_info {
    need_medical_support: true
  }
}

2019-06-20 15:39:31,697 - __main__ - INFO - northstar-service-client: Response :success: true
```


### Building and Running northstar-cloud in Docker

Build docker image.

Run commands from ROOT directory (smart-detective-api)
``` shell
docker build -t northstar-cloud . 
```

Run northstar-cloud's docker container and test APIs 

``` shell
docker images
REPOSITORY                                        TAG                 IMAGE ID            CREATED             SIZE
northstar-cloud                                latest              642ac58bed81        8 hours ago         1.21GB

docker run -d -it --name northstar-cloud northstar-cloud:latest

docker ps
CONTAINER ID        IMAGE                        COMMAND             CREATED             STATUS              PORTS               NAMES
fc9dbd868e87        northstar-cloud:latest   "python3"           4 seconds ago       Up 3 seconds                            northstar-cloud

docker exec -it fc9dbd868e87 bash

CLI:
root@fc9dbd868e87:/usr/src/app# python northstar_cloud/cli/detective_api_cli.py examples/example1.json 
2019-04-30 03:51:34,771 - northstar_cloud.services.detective_api_service - INFO - init_aggregate_witness_events: Initializing input witness events [['fight', 'gunshot', 'fleeing'], ['gunshot', 'falling', 'fleeing']]
2019-04-30 03:51:34,771 - northstar_cloud.services.detective_api_service - INFO - calculate_witness_merge: Calculating witness event merge for: [['fight', 'gunshot', 'fleeing'], ['gunshot', 'falling', 'fleeing']]

detective-api: OUTPUT
detective-api: Witness events given input : [['fight', 'gunshot', 'fleeing'], ['gunshot', 'falling', 'fleeing']] 
detective-api: Witness events prediction      : WITNESS_DECISION_MERGE_ALL_POSSIBLE 
detective-api: Witness events predicted list  : [['fight', 'gunshot', 'falling', 'fleeing']]


gRPC:
root@fc9dbd868e87:/usr/src/app# python northstar_cloud/api/detective_api_rpc_client.py examples/example1.json 
2019-04-30 03:54:09,888 - __main__ - INFO - detective-api-client: Request :witness_events {
  name: "fight"
  name: "gunshot"
  name: "fleeing"
}
witness_events {
  name: "gunshot"
  name: "falling"
  name: "fleeing"
}

2019-04-30 03:54:09,893 - __main__ - INFO - detective-api-client: Response :decision: WITNESS_DECISION_MERGE_ALL_POSSIBLE
witness_events {
  name: "fight"
  name: "gunshot"
  name: "falling"
  name: "fleeing"
}

detective-api: OUTPUT
detective-api: Witness events given input : [['fight', 'gunshot', 'fleeing'], ['gunshot', 'falling', 'fleeing']] 
detective-api: Witness events prediction      : WITNESS_DECISION_MERGE_ALL_POSSIBLE 
detective-api: Witness events predicted list  : [['fight', 'gunshot', 'falling', 'fleeing']]
```


### Building and running northstar-cloud for kubernetes.

Local development is done using [minikube]
First, install kubernetes and minikube and then start minikube cluster.

```
$ minikube start
Starting local Kubernetes v1.6.4 cluster...
Starting VM...
Moving files into cluster...
Setting up certs...
Starting cluster components...
Connecting to cluster...
Setting up kubeconfig...
Kubectl is now configured to use the cluster.
```

```
$ cd northstar-cloud
```

```
$ eval $(minikube docker-env)
```

```
$ docker build -t northstar-cloud .
```

gRPC server port configuration at, 

```bash
~/northstar-cloud/minikube$ls -lrt
total 24
78 Apr 29 15:41 detective-api-config.json
204 Apr 29 22:30 detective-api-config.yaml
1102 Apr 29 22:30 detective-api-deployement.yaml
```

Deploy service on k8s cluster,
```
$ cd minikube/
$ kubectl create -f detective-api-deployement.yaml
```

```bash
$ kubectl get services
NAME                  TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)     AGE
kubernetes            ClusterIP   10.0.0.1     <none>        443/TCP     18d
smart-detective-api   ClusterIP   None         <none>        50051/TCP   9s
```


```bash
$ kubectl get pod
NAME                                   READY     STATUS    RESTARTS   AGE
northstar-cloud-3517672685-n8gdp   1/1       Running   0          43s
```


```bash
kubectl logs smart-detective-api-3517672685-n8gdp
2019-04-30 03:58:35,152 - northstar_cloud.api.detective_api_rpc - INFO - detective-api: service stating...
2019-04-30 03:58:35,156 - northstar_cloud.api.detective_api_rpc - INFO - detective-api: is runnnig at ... localhost:50051
```


```shell
kubectl exec -it smart-detective-api-3517672685-n8gdp bash
root@smart-detective-api-3517672685-n8gdp:/usr/src/app# python northstar_cloud/api/detective_api_rpc_client.py examples/example1.json 
2019-04-30 04:01:48,965 - __main__ - INFO - detective-api-client: Request :witness_events {
  name: "fight"
  name: "gunshot"
  name: "fleeing"
}
witness_events {
  name: "gunshot"
  name: "falling"
  name: "fleeing"
}

2019-04-30 04:01:48,968 - __main__ - INFO - detective-api-client: Response :decision: WITNESS_DECISION_MERGE_ALL_POSSIBLE
witness_events {
  name: "fight"
  name: "gunshot"
  name: "falling"
  name: "fleeing"
}

detective-api: OUTPUT
detective-api: Witness events given input : [['fight', 'gunshot', 'fleeing'], ['gunshot', 'falling', 'fleeing']] 
detective-api: Witness events prediction      : WITNESS_DECISION_MERGE_ALL_POSSIBLE 
detective-api: Witness events predicted list  : [['fight', 'gunshot', 'falling', 'fleeing']]
```

if further questions, please reach me at: hitesh.wadekar@ibm.com


ChangeLog
CHANGES
=======

* norht star cloud mongodb support
* project name and api configuration
* Directory initialization for northstar cloud services
* Initial commit


northstar_cloud/api/detective_api_rpc.py
from concurrent import futures
import time
import os
import grpc

from northstar_cloud.common import logs as logging
from northstar_cloud.api.proto import detective_pb2
from northstar_cloud.api.proto import detective_pb2_grpc
from northstar_cloud.services.\
    northstar_cloud_service import DetectiveApiService as dt_service
from northstar_cloud.common import utils as c_utils

_ONE_DAY_IN_SECONDS = 60 * 60 * 24

LOG = logging.getLogger(__name__)
DEFAULT_FILE_PATH = "minikube/detective-api-config.json"
DEFAULT_BIND = '[::]:50051'
DEFAULT_LOCALHOST_ADDRESS = "localhost:50051"

ROOT_DIR = os.path.abspath(os.curdir)
log_file_path = ROOT_DIR + "/" + DEFAULT_FILE_PATH


class DetectiveServiceServicer(
    detective_pb2_grpc.DetectiveServiceServicer
):
    """Provides methods that implement functionality
    of detective api server."""

    status_to_grpc = {
        c_utils.PossibleEvents.WITNESS_DECISION_CANT_DECIDE:
        detective_pb2.WITNESS_DECISION_CANT_DECIDE,
        c_utils.PossibleEvents.WITNESS_DECISION_PARTIAL_MERGE_POSSIBLE:
        detective_pb2.WITNESS_DECISION_PARTIAL_MERGE_POSSIBLE,
        c_utils.PossibleEvents.WITNESS_DECISION_NO_MERGE_POSSIBLE:
        detective_pb2.WITNESS_DECISION_NO_MERGE_POSSIBLE,
        c_utils.PossibleEvents.WITNESS_DECISION_MERGE_ALL_POSSIBLE:
        detective_pb2.WITNESS_DECISION_MERGE_ALL_POSSIBLE
    }

    def __init__(self):
        self.dt_service = dt_service()

    def _convert_grpc_wtness_req_to_local(self, witness_events):
        local_witness_events = []
        for each_list in witness_events:
            local_witness_events.append(each_list.name)
        return local_witness_events

    def _convert_local_wtness_rpl_to_grpc(self, witness_events):
        req_witness = detective_pb2.GetWitnessMergeReply()
        for each_item in witness_events:
            req_witness.witness_events.extend(
                [detective_pb2.WitnessEvents(name=each_item)]
            )
        return req_witness

    def GetWitnessMergeDecision(self, request, context):
        LOG.info("detetive-api:GetWitnessMergeDecision "
                 "witness-events: %s", request.witness_events)
        # Request recieved.
        witness_events = self._convert_grpc_wtness_req_to_local(
            request.witness_events
        )

        # Service call for decision making.
        self.dt_service.init_aggregate_witness_events(witness_events)
        possible_event_pred, wintess_possible_list = \
            self.dt_service.calculate_witness_merge()

        # Preparing reply
        rply_wintess_possible_merge = \
            self._convert_local_wtness_rpl_to_grpc(wintess_possible_list)
        rply_wintess_possible_merge.decision = \
            self.status_to_grpc[possible_event_pred]
        return rply_wintess_possible_merge


def read_config():
    config = {}
    config = c_utils.read_json_file(log_file_path)
    global DEFAULT_BIND
    global DEFAULT_LOCALHOST_ADDRESS

    if 'bind ' in config.keys():
        DEFAULT_BIND = config['bind']

    if 'default_local_address' in config.keys():
        DEFAULT_LOCALHOST_ADDRESS = config['default_local_address']


def serve():
    read_config()
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=10))
    detective_pb2_grpc.add_DetectiveServiceServicer_to_server(
        DetectiveServiceServicer(), server)
    LOG.info("detective-api: service stating...")

    server.add_insecure_port(DEFAULT_BIND)
    try:
        server.start()
        LOG.info(
            "detective-api: is runnnig "
            "at ... %s" % (DEFAULT_LOCALHOST_ADDRESS)
        )
    except Exception as e:
        LOG.info(
            "detective-api: Failed to "
            "start with exception %s" % (e)
        )
    try:
        while True:
            time.sleep(_ONE_DAY_IN_SECONDS)
    except KeyboardInterrupt:
        LOG.info("detective-api: stopped ")
        server.stop(0)


if __name__ == '__main__':
    serve()
